### PR TITLE
Optimize SqlClient rpc parameter usage

### PR DESCRIPTION
**Pull request by Wraith2**
I have made changes to the _SqlRPC object which change how the system and user parameters are allocated and stored. Originally the system parameters were allocated each execution, I've altered this so the correct number are allocated as needed and re-used cleanly if the query type changes (between prepare, unprepared, proc and exec). 
User parameters are no longer copied in but referenced by index into the parameter collection. 

This is a risky change. It needs careful review to ensure compatibility. 

This PR was split from #32811 and into smaller commits for easier review. It has passed the manual and efcore tests in native mode, the tests cannot be successfully run in managed mode due to #33930 . Performance improvements for this and related PR's are in the original discussion.
/cc @tarikulsabbir,  @AfsanehR @saurabh500 @David-Engel

### CODE REVIEW

**Comment by AfsanehR** [comment]
Thanks for resolving the conflicts @Wraith2 . We will start looking into this pr soon.

**Comment by Wraith2** [comment]
How soon?

**Comment by Wraith2** [comment]
This PR is on page 3, of 3 in the list. It's been open almost 3 months now with no review, it was open for several months before that on the original PR which was deemed too complex. Can I get some feedback or a decision that this change just isn't going to be approved? I keep running into the performance issues this addresses in other areas of the project and having to ignore them knowing I've already fixed it. 
@karelz who do I need to bother to get time allocated?

**Comment by karelz** [comment]
I am guilty of not monitoring & pushing on stale PRs for a while (like 4+ weeks :(), sorry for that.
@AfsanehR @tarikulsabbir @saurabh500 @Gary-Zh  @David-Engel do you have any thoughts on this PR? Feel free to ping me on email if answer is more complicated - I can help! (saying no is also ok with explanation of reasons)

**Comment by karelz** [comment]
@AfsanehR @tarikulsabbir @Gary-Zh any update?

**Comment by tarikulsabbir** [comment]
Hi, Sorry for the delay. I have started looking into this PR.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs [hunk 1]
```diff
@@ -672,17 +678,23 @@ sealed internal class _SqlRPC
         internal int warningsIndexEnd;
         internal SqlErrorCollection warnings;
 
-        internal string GetCommandTextOrRpcName()
+        public SqlParameter GetParameterByIndex(int index, out byte options)
```
Comment (tarikulsabbir) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs]: Do we need to expose this API? can it be internal?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 2]
```diff
@@ -3272,42 +3271,40 @@ private void GetRPCObject(int paramCount, ref _SqlRPC rpc)
...
+            if ((rpc.userParamMap?.Length ?? 0) < userParamCount)
             {
-                for (ii = 0; ii < paramCount; ii++)
-                    rpc.paramoptions[ii] = 0;
+                Array.Resize(ref rpc.userParamMap, userParamCount);
             }
         }
 
+
         private void SetUpRPCParameters(_SqlRPC rpc, int startCount, bool inSchema, SqlParameterCollection parameters)
```
Comment (tarikulsabbir) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: startCount parameter is no longer in use. Can it be removed?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 3]
```diff
@@ -3272,42 +3271,40 @@ private void GetRPCObject(int paramCount, ref _SqlRPC rpc)
...
-            else
+            if ((rpc.userParamMap?.Length ?? 0) < userParamCount)
             {
-                for (ii = 0; ii < paramCount; ii++)
-                    rpc.paramoptions[ii] = 0;
+                Array.Resize(ref rpc.userParamMap, userParamCount);
             }
         }
 
+
```
Comment (tarikulsabbir) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Nit: Extra space

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs (hunk 1 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs]: Its public on an internal class which makes it effectively internal, but I've changed it to internal anyway.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 3 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: removed.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 2 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: yup, done.

**Comment by karelz** [comment]
@tarikulsabbir @Wraith2 is it blocked only on passing tests?

**Comment by Wraith2** [comment]
As far as I know, just merged to master which will re-run the CI.

**Comment by tarikulsabbir** [comment]
> @tarikulsabbir @Wraith2 is it blocked only on passing tests?

Hi @karelz and @Wraith2 , we are currently running EFCore tests on different platforms. We had a few issues on our test environment which is now resolved. Will update again soon.

**Comment by Wraith2** [comment]
CI failures are valuetuple and linq, unrelated to this change.

**Comment by tarikulsabbir** [comment]
The tests are green. There were a few environmental issues which caused the delay. Anyway I will merge the PR now. Thanks for your patience.

**Pull request merged**